### ISSUE TITLE
A performance issue with morphism detection

### ISSUE THREAD

**Issue by pi8027** [math-comp/algebra-tactics#27]
```coq
Goal forall (F : numFieldType),
let x : F := 2%:~R in
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x =
4%:~R ^+ 100.
Proof.
move=> F /=.
Time by field.
(*
Reification: 0.019452 sec.
Reflection: 0.197635 sec.
Finished transaction in 0.246 secs (0.246u,0.s) (successful)
*)
Qed.

Goal forall (F : numFieldType),
let x : F := ratr 2%:~R in (* ratr is a ring morphism *)
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x *
x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x * x =
4%:~R ^+ 100.
Proof.
move=> F /=.
Time by field.
(*
Reification: 0.221024 sec.
Reflection: 0.251970 sec.
Finished transaction in 0.544 secs (0.544u,0.s) (successful)
*)
Qed.
```

**Comment by pi8027** [comment]
I happened to see another performance issue in the example above, which has been fixed in #29. The performance issue of morphism detection is not easy to fix. I attempted to do that by implementing a restricted version of canonical structure inference in Elpi and it does not improve anything.

**Comment by pi8027** [comment]
I confirm that this performance issue has disappeared.
